statsprinter.go

```diff
@@ -370,141 +317,210 @@ type JSONData struct {
-		data.Success = &t
+func (p *jsonPrinter) printProbeFail(hostname, ip string, port uint16, streak uint) {
+	var (
+		// for *bool fields
+		f    = false
+		t    = true
+		data = JSONData{
+			Type:                    probeEvent,
+			Hostname:                hostname,
+			Addr:                    ip,
+			Port:                    port,
+			IsIP:                    &t,
+			Success:                 &f,
+			TotalUnsuccessfulProbes: streak,
+		}
+	)
+
+	if hostname != "" {
+		data.IsIP = &f
+		data.Message = fmt.Sprintf("No reply from %s (%s) on port %d",
+			hostname, ip, port)
 	} else {
-		data.TotalUnsuccessfulProbes = j.totalUnsuccessfulProbes
+		data.Message = fmt.Sprintf("No reply from %s on port %d",
+			ip, port)
 	}
 
-	_ = printJson(data)
+	p.print(data)
 }
 
 // printStatistics prints all gathered stats when program exits.
-func (j *statsJsonPrinter) printStatistics() {
+func (p *jsonPrinter) printStatistics(s stats) {
 	data := JSONData{
-		Type:      statsEvent,
-		Message:   fmt.Sprintf("stats for %s", j.hostname),
-		Hostname:  j.hostname,
-		Timestamp: time.Now(),
-
-		StartTimestamp:          &j.startTime,
-		TotalDowntime:           j.totalDowntime.Seconds(),
-		TotalPackets:            j.totalSuccessfulProbes + j.totalUnsuccessfulProbes,
-		TotalSuccessfulProbes:   j.totalSuccessfulProbes,
-		TotalUnsuccessfulProbes: j.totalUnsuccessfulProbes,
-		TotalUptime:             j.totalUptime.Seconds(),
+		Type:     statisticsEvent,
+		Message:  fmt.Sprintf("stats for %s", s.hostname),
+		Hostname: s.hostname,
+
+		StartTimestamp:          &s.startTime,
+		TotalDowntime:           s.totalDowntime.Seconds(),
+		TotalPackets:            s.totalSuccessfulProbes + s.totalUnsuccessfulProbes,
+		TotalSuccessfulProbes:   s.totalSuccessfulProbes,
+		TotalUnsuccessfulProbes: s.totalUnsuccessfulProbes,
+		TotalUptime:             s.totalUptime.Seconds(),
 	}
 
 	data.TotalPacketLoss = (float32(data.TotalUnsuccessfulProbes) / float32(data.TotalPackets)) * 100
 
-	if !j.lastSuccessfulProbe.IsZero() {
-		data.LastSuccessfulProbe = &j.lastSuccessfulProbe
-	}
-	if !j.lastUnsuccessfulProbe.IsZero() {
-		data.LastUnsuccessfulProbe = &j.lastUnsuccessfulProbe
+	if !s.lastSuccessfulProbe.IsZero() {
+		data.LastSuccessfulProbe = &s.lastSuccessfulProbe
 	}
-
-	/* calculate the last longest time */
-	if !j.wasDown {
-		calcLongestUptime(j.stats,
-			time.Duration(j.ongoingSuccessfulProbes)*time.Second)
-	} else {
-		calcLongestDowntime(j.stats,
-			time.Duration(j.ongoingUnsuccessfulProbes)*time.Second)
+	if !s.lastUnsuccessfulProbe.IsZero() {
+		data.LastUnsuccessfulProbe = &s.lastUnsuccessfulProbe
 	}
 
-	if j.longestUptime.duration != 0 {
-		data.LongestUptime = fmt.Sprintf("%.3f", j.longestUptime.duration.Seconds())
-		data.LongestUptimeStart = &j.longestUptime.start
-		data.LongestUptimeEnd = &j.longestUptime.end
+	if s.longestUptime.duration != 0 {
+		data.LongestUptime = fmt.Sprintf("%.3f", s.longestUptime.duration.Seconds())
```

Comment:
This should not be a float.

Reply:
I bit confused about this one

We can't pass `%d`, because `.Seconds()` is float64, so you want me to simply change it to `%.0f`?

Reply:
or we can type cast it to `int`? Because on the output it shows something like:

```console
...
"total_duration": "3.000",
"longest_downtime": "3.000",
...
```

Reply:
I'd go with `%.0f` (should be just `3`, not `3.0` in the output) if you don't mind then. When custom timeout will be introduced we should probably update to `%.1f` or something, but for now it'll do.

I don't really want to do a type-conversion, because we only need this info for output

Reply:
Correct.

Just to make myself clear, we display fractions when necessary. For instance, the `min/avg/max` values. However, in case of `total_duration` or anything of that nature, which translates to seconds, we do not need to display the fractions, because seconds have no decimal points or fractions.